src/Chainweb/Miner/Coordinator.hs

```diff
@@ -134,82 +163,44 @@ newtype PrevTime = PrevTime BlockCreationTime
     -> ChainChoice
-    -> MinerStatus
+    -> Miner
     -> WebBlockHeaderDb
         -- ^ this is used to lookup parent headers that are not in the cut
         -- itself.
     -> PactExecutionService
     -> TVar PrimedWork
     -> Cut
     -> IO (T2 WorkHeader PayloadData)
-newWork logFun choice eminer hdb pact tpw c = do
+newWork logFun choice eminer@(Miner mid _) hdb pact tpw c = do
 
     -- Randomly pick a chain to mine on, unless the caller specified a specific
     -- one.
     --
     cid <- chainChoice c choice
     logFun @T.Text Debug $ "newWork: picked chain " <> sshow cid
 
-    mr <- case eminer of
-        Primed m -> primed m cid <$> readTVarIO tpw
-        Plebian m -> do
-            logFun @T.Text Debug
-                $ "newWork: call getCutExtension for chain "
-                <> sshow cid <> " on cut " <> cutIdToTextShort (_cutId c)
-            public m cid
+    PrimedWork pw <- readTVarIO tpw
+    let mr = T2
+            <$> join (HM.lookup mid pw >>= HM.lookup cid)
+            <*> getCutExtension c cid
+
     case mr of
         Nothing -> do
             logFun @T.Text Debug $ "newWork: chain " <> sshow cid <> " not mineable"
             newWork logFun (TriedLast cid) eminer hdb pact tpw c
-        Just (T2 payload extension) -> do
-            let !phash = _payloadDataPayloadHash payload
-            !wh <- newWorkHeader hdb extension phash
-            pure $ T2 wh payload
-  where
-    primed
-        :: Miner
-        -> ChainId
-        -> PrimedWork
-        -> Maybe (T2 PayloadData CutExtension)
-    primed (Miner mid _) cid (PrimedWork pw) = T2
-        <$> join (HM.lookup mid pw >>= HM.lookup cid)
-        <*> getCutExtension c cid
-
-    public
-        :: Miner
-        -> ChainId
-        -> IO (Maybe (T2 PayloadData CutExtension))
-    public miner cid = case getCutExtension c cid of
-        Nothing -> do
-            logFun @T.Text Debug
-                $ "newWork.public: failed to get adjacent parents."
-                <> " Parent: " <> encodeToText (ObjectEncoded $ c ^?! ixg cid)
-                <> " Cuthashes: " <> encodeToText (cutToCutHashes Nothing c)
-            pure Nothing
-        Just ext -> do
-            logFun @T.Text Debug
-                $ "newWork.public: got extension."
-                <> " Parent: " <> encodeToText (ObjectEncoded $ _parentHeader $ _cutExtensionParent ext)
-                <> " Adjacents: " <> encodeToText (_cutExtensionAdjacentHashes ext)
-                <> " Cuthashes: " <> encodeToText (cutToCutHashes Nothing (_cutExtensionCut ext))
-            -- This is an expensive call --
-            payload <- trace logFun "Chainweb.Miner.Coordinator.newWork.newBlock" () 1
-                (_pactNewBlock pact miner $ _cutExtensionParent ext)
-            pure . Just $ T2 (payloadWithOutputsToPayloadData payload) ext
+        Just (T2 (payload, parent) extension)
+            | parent == _blockHash (_parentHeader (_cutExtensionParent extension)) -> do
```

Comment:
Adding this check fixes a race where a new cut is picked up before the primed work is updated. I this case it could happen that that a mining client would receive an outdated payload. The respective block would only fail later during validation in the cut pipeline.